nixos/network: add gre virtual interfaces

hexagonal-sun opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Add options that allow `gre` tunnels to be creating in the nixOS config.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

pennae commented on nixos/modules/tasks/network-interfaces-scripted.nix:
```diff
@@ -253,7 +253,11 @@ let
           };
 
-        createTunDevice = i: nameValuePair "${i.name}-netdev"
+          createTunDevice = i: nameValuePair "${i.name}-netdev"
```
```suggestion
        createTunDevice = i: nameValuePair "${i.name}-netdev"
```
did your editor indent this by mistake?

hexagonal-sun replied:
Whops. Yeah my editor added the indent.  Removed.

pennae commented on nixos/modules/tasks/network-interfaces-scripted.nix:
```diff
@@ -253,7 +253,11 @@ let
 
-        createTunDevice = i: nameValuePair "${i.name}-netdev"
+          createTunDevice = i: nameValuePair "${i.name}-netdev"
+          ( let
+              greType = if i.virtualType == "gretun" then "gre" else i.virtualType;
+              greLocal = optionalString (i.gre.local != null) " local ${i.gre.local}";
```
this entire block could be moved to the `script` attribute that uses it. would probably be easier to read like that, adding lets like this is rather uncommon

hexagonal-sun replied:
Yep, that clears things up. Done.

pennae reviewed (changes requested):
this also needs to be added to the networkd backend and tested for both backends

hexagonal-sun commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -1215,6 +1230,12 @@ in
         '';
+      })) ++ (forEach interfaces (i: {
+        assertion = (i.virtual && (i.virtualType == "gretun" ||
+                                   i.virtualType == "gretap")) -> i.gre.remote != null;
```
Agreed.  Changed `gre.remote` to be `nullOr str`.

hexagonal-sun commented:
> this also needs to be added to the networkd backend and tested for both backends

Added systemd back-end and tested with both.  Some small changes I've made atop of your comments:

 - I've renamed `gretun` to `gre`.  This is consistent with how other tools call this mode and it claens up the code from translating `gretun` -> `gre`.
 - I'm using `hasPrefix "gre"` to check whether the virtual type is GRE.

pennae commented:
> Added systemd back-end and tested with both.

can you add a nixos test for the testing you've done as well? would be nice to ensure this keeps working even when other things are changed. apart from that, looks good to me :+1:

hexagonal-sun commented:
@pennae Thanks for taking a look but, sorry, I decided to do a bit of a rewrite and ask for another review.  `gre` interfaces are more like `sit` interfaces, i.e. they're a point-to-point link which don't have an IP address, netmask, etc.  The previous configuration would allow the user to set an IP address on the GRE tunnel which doesn't make sense.  The new rewrite should fix this.  I'll also look at creating some tests.

Thanks!

pennae commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -996,6 +997,64 @@ in
+        }
+      '';
+      description = ''
+        This option allows you to define a Generic Routing Encapsulation (GRE) tunnel.
```
```suggestion
        This option allows you to define Generic Routing Encapsulation (GRE) tunnels.
```

pennae commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -996,6 +997,64 @@ in
       });
     };
 
+    networking.gre = mkOption {
```
in keeping with other options this should be pluralized, but `gres` seems more confusing than helpful at first glance. how about
```suggestion
    networking.greTunnels = mkOption {
```

pennae commented on nixos/modules/tasks/network-interfaces-scripted.nix:
```diff
@@ -532,6 +532,34 @@ let
+        createGreDevice = n: v: nameValuePair "${n}-netdev"
+          (let
+            deps = deviceDependency v.dev;
+            greType = if v.type == "tun" then "gre" else "gretap";
```
this expression is duplicated a few times. you could add it to the option definition to have the option checked against the declared type, but stored for later use as though this expression had always been applied to it first.
```nix
  apply = v: if v == "tun" then "gre" else "gretap";
```

hexagonal-sun replied:
Nice!  I didn't know about this option.

hexagonal-sun commented:
> > Added systemd back-end and tested with both.
> 
> can you add a nixos test for the testing you've done as well? would be nice to ensure this keeps working even when other things are changed. apart from that, looks good to me 👍

I've now added a gre test case which uses a GRE tunnel to get layer2 packets across VLANs.  Hope that's okay?

pennae commented on nixos/tests/networking.nix:
```diff
@@ -101,9 +101,6 @@ let
             { address = "192.168.1.3"; prefixLength = 32; }
             { address = "192.168.1.10"; prefixLength = 32; }
           ];
-          interfaces.eth2.ipv4.addresses = mkOverride 0 [
-            { address = "192.168.2.2"; prefixLength = 24; }
-          ];
```
is this an intentional change? removing this breaks the test.

hexagonal-sun replied:
Oops. Sorry that must have been while I was figuring out how the tests worked!  Should be reinstated now.

pennae reviewed (commented):
looking good! the test runs with both scripted and networkd setups, but looks like an unrelated change broke the static test. other than that, seems good to me.

hexagonal-sun committed:
nixos/networking: add options for configuring a GRE tunnel

Add `networking.greTunnels` option that allows a GRE tunnel to be
configured in NixOS.

hexagonal-sun commented:
> looking good! the test runs with both scripted and networkd setups, but looks like an unrelated change broke the static test. other than that, seems good to me.

Great!  The `static` test should now be fixed.

hexagonal-sun commented:
@tomberek Are you okay to review please? If you're happy, are okay to merge this? Thanks! :-)

hexagonal-sun commented:
@SuperSandro2000 If you're happy with the above, is it okay to get merged?  Thanks!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/666

symphorien commented:
@GrahamcOfBorg test networking
